### PR TITLE
Shadow control for zones, models and shapes

### PR DESCRIPTION
**Pull request by NissimHadar**
# General
This PR enables content creator control of casting shadows.  
1.  Shadows may be turned on/off per zone
2.  Shadows may be turned on/off per entity/shape.
# Requirements
1. Master shadow flag defaults to on (Menu->Developer->Graphics...) DONE
2. Add a shadow cast flag to the keylight component of the zone, defaulting to off DONE
3. Do not run shadow task when shadows are not active.
4. Add a shadow cast flag to Entity, dafaulting to on.  Entities will cast shadows iff the master & active zone & entity flags are all active DONE
5. Avatars will always cast shadows (provided that the master & the active zone flags are on DONE
6. Shadows/Ambient Occlusion global control - use menu checkboxes instead of qml screen.
# Testing
Tests are defined in: https://highfidelity.testrail.net/index.php?/suites/view/755&group_by=cases:title&group_order=asc

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7205/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7205/HighFidelity-Beta-PR12407-a92765a83a3578572c1ce9951c807d417b6744f5.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7205/HighFidelity-Beta-PR12407-a92765a83a3578572c1ce9951c807d417b6744f5.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7208/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7208/HighFidelity-Beta-PR12407-adb02d69f9c50721d9e186870ae8ca6cf201a413.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7208/HighFidelity-Beta-PR12407-adb02d69f9c50721d9e186870ae8ca6cf201a413.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7210/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7210/HighFidelity-Beta-PR12407-7e99570824418cc1a57b72cad44ddd8c386a1bff.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7210/HighFidelity-Beta-PR12407-7e99570824418cc1a57b72cad44ddd8c386a1bff.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7235/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7235/HighFidelity-Beta-PR12407-ff43e4c2ad45dcc9ab8e75771c6a8c4905c79753.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7235/HighFidelity-Beta-PR12407-ff43e4c2ad45dcc9ab8e75771c6a8c4905c79753.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7243/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7243/HighFidelity-Beta-PR12407-72e89851f3c2936bbf3da1f065fb3bc4fb4b919e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7243/HighFidelity-Beta-PR12407-72e89851f3c2936bbf3da1f065fb3bc4fb4b919e.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7272/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7275/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7275/HighFidelity-Beta-PR12407-47924a58f9732e62ef79efa065bd944781193ebd.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7275/HighFidelity-Beta-PR12407-47924a58f9732e62ef79efa065bd944781193ebd.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7298/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7298/HighFidelity-Beta-PR12407-4d4e3d6aaed1dd50bffc1ec52c306429ae187031.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7298/HighFidelity-Beta-PR12407-4d4e3d6aaed1dd50bffc1ec52c306429ae187031.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7300/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7300/HighFidelity-Beta-PR12407-7ba8a9935a5d18c2448e203c3556f1bd59ddbab2.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7300/HighFidelity-Beta-PR12407-7ba8a9935a5d18c2448e203c3556f1bd59ddbab2.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7308/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7310/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7310/HighFidelity-Beta-PR12407-b2a06cdb98d52fbb34cb45bd7031b22393a357b0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7310/HighFidelity-Beta-PR12407-b2a06cdb98d52fbb34cb45bd7031b22393a357b0.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7346/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7346/HighFidelity-Beta-PR12407-af484d0af6a5793671793104cf24672c788cb39c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7346/HighFidelity-Beta-PR12407-af484d0af6a5793671793104cf24672c788cb39c.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7359/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7359/HighFidelity-Beta-PR12407-b519ebd6f1de42ff496edb578167a1e06b2ee56c.dmg

File: libraries/entities/src/KeyLightPropertyGroup.cpp [hunk 1]
```diff
@@ -64,26 +70,32 @@ void KeyLightPropertyGroup::listChangedProperties(QList<QString>& out) {
...
+    OctreeElement::AppendState& appendState) const 
+{
 
     bool successPropertyFits = true;
     
     APPEND_ENTITY_PROPERTY(PROP_KEYLIGHT_COLOR, getColor());
     APPEND_ENTITY_PROPERTY(PROP_KEYLIGHT_INTENSITY, getIntensity());
     APPEND_ENTITY_PROPERTY(PROP_KEYLIGHT_DIRECTION, getDirection());
-    
+    APPEND_ENTITY_PROPERTY(PROP_KEYLIGHT_CAST_SHADOWS, getCastShadows());
```
Comment (samcake) [inline libraries/entities/src/KeyLightPropertyGroup.cpp]: I would suggest to use the property name CAST_SHADOW
WIthout an S

File: libraries/entities/src/EntityPropertyFlags.h [hunk 2]
```diff
@@ -20,6 +20,7 @@ enum EntityPropertyList {
 
     // these properties are supported by the EntityItem base class
     PROP_VISIBLE,
+    PROP_CAN_CAST_SHADOW,
```
Comment (samcake) [inline libraries/entities/src/EntityPropertyFlags.h]: YOu could get away with either:
CAST_SHADOW
or
SHADOW_CASTER

File: libraries/entities/src/EntityItemPropertiesDefaults.h [hunk 3]
```diff
@@ -46,6 +46,8 @@ const quint32 ENTITY_ITEM_DEFAULT_STATIC_CERTIFICATE_VERSION = 0;
 const float ENTITY_ITEM_DEFAULT_ALPHA = 1.0f;
 const float ENTITY_ITEM_DEFAULT_LOCAL_RENDER_ALPHA = 1.0f;
 const bool ENTITY_ITEM_DEFAULT_VISIBLE = true;
+const bool ENTITY_ITEM_DEFAULT_CAST_SHADOWS { false };
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityItemPropertiesDefaults.h]: `ENTITY_ITEM_DEFAULT_CAST_SHADOWS` has been added but is never used.

File: libraries/entities/src/EntityPropertyFlags.h [hunk 4]
```diff
@@ -205,6 +206,11 @@ enum EntityPropertyList {
 
     PROP_HAZE_MODE,
 
+    PROP_KEYLIGHT_COLOR,
+    PROP_KEYLIGHT_INTENSITY,
+    PROP_KEYLIGHT_DIRECTION,
+    PROP_KEYLIGHT_CAST_SHADOWS,
+
```
Comment (AndrewMeadows) [inline libraries/entities/src/EntityPropertyFlags.h]: I believe **NEW** properties **MUST** be added at the bottom.  Else terrible things happen, or maybe terrible things USED to happen back when we stored entity-server state as binary packets.  I don't fully understand the current state of this gotcha.  Perhaps @zappoman or someone else knows?

Also, are we sure we want to move the old, previously recycled, properties into their own enum value?  There is overhead for every **new** entity property we add.  Is it worth it for these properties?

File: libraries/entities/src/KeyLightPropertyGroup.cpp [hunk 5]
```diff
@@ -64,26 +70,32 @@ void KeyLightPropertyGroup::listChangedProperties(QList<QString>& out) {
...
+
     return true;
 }
 
 bool KeyLightPropertyGroup::decodeFromEditPacket(EntityPropertyFlags& propertyFlags, const unsigned char*& dataAt, 
-    int& processedBytes) {
+    int& processedBytes) 
+{
```
Comment (AndrewMeadows) [inline libraries/entities/src/KeyLightPropertyGroup.cpp]: Incorrect formatting here for the opening brace of a mundane class method to have its own line, as per the coding standard [section 4.1.4](https://github.com/highfidelity/hifi/wiki/Coding-Standard#414--method-definitions-should-have-the-following-form).

File: scripts/system/html/entityProperties.html [hunk 6]
```diff
@@ -532,6 +530,10 @@
                         <div></div>
                     </div>
                 </div>
+				<div class="zone-group zone-section keylight-section property checkbox">
+					<input type="checkbox" id="property-zone-key-light-cast-shadows">
+					<label for="property-zone-key-light-cast-shadows">Cast Shadows</label>
+				</div>
```
Comment (AndrewMeadows) [inline scripts/system/html/entityProperties.html]: These lines are indented with tab characters, but should use spaces.

File: interface/resources/qml/hifi/dialogs/GraphicsPreferencesDialog.qml [hunk 7]
```diff
@@ -1,19 +0,0 @@
-import QtQuick 2.5
-import Qt.labs.settings 1.0
-
-import "../../dialogs"
-
-PreferencesDialog {
```
Comment (ZappoMan) [inline interface/resources/qml/hifi/dialogs/GraphicsPreferencesDialog.qml]: did you really mean to remove this? this UI has the ambient occlusion setting on it as well as shadows, I know you probably meant to remove shadows but the whole dialog?

File: interface/src/avatar/MyAvatar.cpp [hunk 8]
```diff
@@ -1819,12 +1821,6 @@ void MyAvatar::attach(const QString& modelURL, const QString& jointName,
     Avatar::attach(modelURL, jointName, translation, rotation, scale, isSoft, allowDuplicates, useSaved);
 }
 
-void MyAvatar::setVisibleInSceneIfReady(Model* model, const render::ScenePointer& scene, bool visible) {
```
Comment (ZappoMan) [inline interface/src/avatar/MyAvatar.cpp]: did you really mean to remove this?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7427/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7427/HighFidelity-Beta-PR12407-bb04f3f53328906baa542c16f8f6f9d1a22f249a.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7427/HighFidelity-Beta-PR12407-bb04f3f53328906baa542c16f8f6f9d1a22f249a.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7442/app_12407_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7442/HighFidelity-Beta-PR12407-3150b963c0106f163ea5aec03a34a23c5220dda3.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7442/HighFidelity-Beta-PR12407-3150b963c0106f163ea5aec03a34a23c5220dda3.exe

File: interface/resources/qml/hifi/dialogs/GraphicsPreferencesDialog.qml (hunk 7 above)
Comment (NissimHadar) [reply interface/resources/qml/hifi/dialogs/GraphicsPreferencesDialog.qml]: Yes.  The Ambient Occlusion setting has also been added to the menu.

File: interface/src/avatar/MyAvatar.cpp (hunk 8 above)
Comment (NissimHadar) [reply interface/src/avatar/MyAvatar.cpp]: Yes.  This method is never called.  I wasted some time on  a breakpoint there...

**Comment by qlimnavarro** [comment]
Pass. @conklin94122

**Comment by tapalisa** [comment]
Someone using this has not set their tester flag. Please check your settings.

**Pull request merged**